- add jwt token auth;

Miller1995 opened the pull request:
- in pom.xml add new dependency java-jwt;
- change the logic in SecurityFilterChain in class SecurityConfig (delete login and logout form/url, change url for permit all and add new role SUPER_ADMIN);
- in SpringConfig add new bean AuthenticationManager for authentication process;
- change UserAuthDTO (delete column dateOfBirth);
- add column user_email in UserAuthEntity;
- add new class AuthResponse for jwt-token response;
- add new interface AuthService;
- add new service class AuthServiceImpl, in this class implements 2 methods for register and authenticate user;
- add new class JWTAuthFilter, in this class build the logic for jwt-token authorization;
- add new class JWTUtil, in this class build the logic for generateToken and validateToken;
- add new dto class userRegisterDTO for register user in database;
- in AuthController created 2 endpoints for register and login user;

Miller1995 committed:
- add jwt token auth;
- in pom.xml add new dependency java-jwt;
- change the logic in SecurityFilterChain in class SecurityConfig (delete login and logout form/url, change url for permit all and add new role SUPER_ADMIN);
- in SpringConfig add new bean AuthenticationManager for authentication process;
- change UserAuthDTO (delete column dateOfBirth);
- add column user_email in UserAuthEntity;
- add new class AuthResponse for jwt-token response;
- add new interface AuthService;
- add new service class AuthServiceImpl, in this class implements 2 methods for register and authenticate user;
- add new class JWTAuthFilter, in this class build the logic for jwt-token authorization;
- add new class JWTUtil, in this class build the logic for generateToken and validateToken;
- add new dto class userRegisterDTO for register user in database;
- in AuthController created 2 endpoints for register and login user;

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/configs/JWTAuthFilter.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Override
+    protected void doFilterInternal( @NonNull HttpServletRequest request,
```
Can this method be simplified? Can we maybe extract some logic in smaller methods?
In this way, we can improve visibility and make it easier to understand.

Miller1995 replied:
ok

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/configs/JWTAuthFilter.java:
```diff
@@ -0,0 +1,66 @@
+            jwt = authHeader.substring(7);
+
+            if (jwt.isBlank()){
+                response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid JWT Token in Bearer Header");
```
As advice, you should not specify too many details about an exception, because this can lead to important leakage of information to a potential attacker. 
`Invalid JWT Token` I would say it's more than enough as a message exception.

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/controllers/AuthController.java:
```diff
@@ -1,36 +1,37 @@
+    @Operation(summary = "this endpoint is used to register a new user in database")
+    public ResponseEntity<UserRegisterDTO> performRegistration(@RequestBody UserRegisterDTO userRegisterDTO){
+        authService.register(userRegisterDTO);
+        return ResponseEntity.ok(userRegisterDTO);
```
Usually, you want to have a service method that is used with a `POST` HTTP action that returns a DTO as a result of the save to the database.
In our case, we could have `return ResponseEntity.ok( authService.register(userRegisterDTO));`
We could have a data mismatch if we return just the DTO from the request body as a response. Instead, we should return the response from the service layer.

Miller1995 replied:
Done!

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/controllers/AuthController.java:
```diff
@@ -1,36 +1,37 @@
+    @Operation(summary = "this endpoint is used to login the user in app and return the jwt token for this user")
+    public ResponseEntity<AuthResponse> performAuthenticate(@RequestBody UserAuthDTO userAuthDTO){
+        AuthResponse authResponse = authService.authenticate(userAuthDTO);
+        return ResponseEntity.ok(authResponse);
```
You can have `return ResponseEntity.ok(authService.authenticate(userAuthDTO));`, and save one line of code.

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/securities/JWTUtil.java:
```diff
@@ -0,0 +1,44 @@
+                .sign(Algorithm.HMAC256(jwtSecret));
+    }
+
+
```
One empty row.

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/services/auth/AuthService.java:
```diff
@@ -0,0 +1,28 @@
+     *
+     * @param userRegisterDTO
+     */
+    void register(UserRegisterDTO userRegisterDTO);
```
As I was saying in the controller, would this method help us if we return `UserRegisterDTO`?

AlexIliescu11 commented on src/main/java/md/miller1995/Dealership/services/auth/AuthServiceImpl.java:
```diff
@@ -0,0 +1,88 @@
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@Transactional(readOnly = true)
```
If you have @Transactional on the method, it will give us any advantage to use it also in the class?

Miller1995 replied:
@AlexIliescu11 If I use @Transactional (readOnly=true) on the class, in this case I don't protect my class from some accidental manipulations in the database?               I mean, for the Data Manipulation Language, in this case, it will be necessary to add the additional annotation @Transactional that permit write, on the methods.            Or did I understand something wrong?

Miller1995 committed:
- delete DealershipApplicationTests.java;
- change some things and logic in next classes ( AuthController, AuthService, AuthServiceImpl, JWTAuthFilter and JWTUtil) like delete empty rows, add some logic for return type etc.
